clang/lib/Sema/SemaChecking.cpp

```diff
@@ -179,6 +179,56 @@ static bool SemaBuiltinCHERICapCreate(Sema &S, CallExpr *TheCall) {
   return false;
 }
 
+static bool checkBuiltinArgument(Sema &S, CallExpr *E, unsigned ArgIndex);
+
+static bool checkCapArg(Sema &S, CallExpr *TheCall, unsigned ArgIndex,
+                        QualType &SrcTy) {
+  clang::Expr *Source = TheCall->getArg(ArgIndex);
+  SrcTy = Source->getType();
+  // We should also be able to use it with arrays (TODO: and functions?).
+  if (SrcTy->canDecayToPointerType() && SrcTy->isArrayType()) {
+    SrcTy = S.Context.getDecayedType(SrcTy);
+  }
+  if (!SrcTy->isCapabilityPointerType() && !SrcTy->isIntCapType()) {
+    S.Diag(Source->getExprLoc(), diag::err_typecheck_expect_capability_operand)
+        << SrcTy;
+    return true;
+  }
+  ExprResult SrcArg = S.PerformCopyInitialization(
+      InitializedEntity::InitializeParameter(S.Context, SrcTy, false),
+      SourceLocation(), Source);
+  if (SrcArg.isInvalid())
+    return true;
+  TheCall->setArg(ArgIndex, SrcArg.get());
+  return false;
+}
+
+/// The CHERI accessors should accept both pointer types and (u)intcap_t.
+static bool SemaBuiltinCheriGet(Sema &S, CallExpr *TheCall) {
+  if (checkArgCount(S, TheCall, 1))
+    return true;
+  QualType SrcTy;
+  if (checkCapArg(S, TheCall, 0, SrcTy))
+    return true;
+  return false;
+}
+
+/// The result type of CHERI mutators should accept both pointer types and
+/// uintcap_t and have the same (const-preserving) result type.
+static bool SemaBuiltinCheriSet(Sema &S, CallExpr *TheCall) {
```

Comment:
Inline for same reasons as SemaBuiltinCheriGet? (assuming you also agree with my comments about either de-duplicating the case statement bodies or making them polymorphic in their second arguments and thus different anyway)